internal/expressions/function/functions.go

```diff
@@ -374,12 +443,77 @@ func getCommit(
 	}
 }
 
-// getImage returns a function that finds container images based on repository
+// getCommitFromWarehouse returns a function that finds Git commits based on repository URL.
+func getCommitFromWarehouse(
+	ctx context.Context,
+	wh *kargoapi.Warehouse,
+	artifacts *kargoapi.DiscoveredArtifacts,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) != 1 {
+			return nil, fmt.Errorf("expected 1 argument, got %d", len(a))
+		}
+
+		repoURL, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("first argument must be string, got %T", a[0])
+		}
+
+		logger := logging.LoggerFromContext(ctx).WithValues(
+			"repoURL", repoURL,
+			"warehouse", wh.Name,
+		)
+
+		var latestCommit *kargoapi.DiscoveredCommit
+		for _, s := range wh.Spec.Subscriptions {
+			if s.Git != nil && s.Git.RepoURL == repoURL && len(artifacts.Git) != 0 {
```

Comment:
These URLs should be normalized before comparison. Same goes for the other functions.